gunicorn/http/message.py

```diff
@@ -180,97 +180,103 @@ def __init__(self, cfg, unreader, peer_addr, req_number=1):
 
+        # get request line
+        line = self.read_line(unreader, buffer, self.limit_request_line)
         self.parse_request_line(line)
-        buf = io.BytesIO()
-        buf.write(rbuf)
 
         # Headers
-        data = buf.getvalue()
-        idx = data.find(b"\r\n\r\n")
-
-        done = data[:2] == b"\r\n"
         while True:
-            idx = data.find(b"\r\n\r\n")
-            done = data[:2] == b"\r\n"
+            idx = buffer.find(b"\r\n\r\n")
+            done = buffer[:2] == b"\r\n"
 
             if idx < 0 and not done:
-                self.get_data(unreader, buf)
-                data = buf.getvalue()
-                if len(data) > self.max_buffer_headers:
+                self.get_data(unreader, buffer)
+                if len(buffer) > self.max_buffer_headers:
                     raise LimitRequestHeaders("max buffer headers")
             else:
                 break
 
         if done:
-            self.unreader.unread(data[2:])
+            self.unreader.unread(buffer[2:])
             return b""
 
-        self.headers = self.parse_headers(data[:idx])
+        self.headers = self.parse_headers(buffer[:idx])
 
-        ret = data[idx + 4:]
-        buf = None
+        # Body
+        ret = buffer[idx + 4:]
+        buffer = None
         return ret
 
-    def read_line(self, unreader, buf, limit=0):
-        data = buf.getvalue()
-
+    def read_line(self, unreader, buffer, limit=0):
         while True:
-            idx = data.find(b"\r\n")
+            idx = buffer.find(b"\r\n")
             if idx >= 0:
                 # check if the request line is too large
                 if idx > limit > 0:
                     raise LimitRequestLine(idx, limit)
                 break
-            if len(data) - 2 > limit > 0:
-                raise LimitRequestLine(len(data), limit)
-            self.get_data(unreader, buf)
-            data = buf.getvalue()
+            if len(buffer) - 2 > limit > 0:
+                raise LimitRequestLine(len(buffer), limit)
+            self.get_data(unreader, buffer)
 
-        return (data[:idx],  # request line,
-                data[idx + 2:])  # residue in the buffer, skip \r\n
+        result = buffer[:idx]   # request line
+        buffer[:] = buffer[idx + 2:]  # residue in the buffer, skip \r\n
+        return result
 
-    def proxy_protocol(self, line):
+    def read_bytes(self, unreader, buffer, size):
+        while True:
+            bytes_read = len(buffer)
+            if bytes_read >= size:
+                break
+            self.get_data(unreader, buffer)
+
+        result, buffer[:] = buffer[:size], buffer[size:]
+        return result
+
+    def proxy_protocol(self, unreader, buffer):
         """\
         Detect, check and parse proxy protocol.
 
         :raises: ForbiddenProxyRequest, InvalidProxyLine.
         :return: True for proxy protocol line else False
         """
         if not self.cfg.proxy_protocol:
-            return False
+            return
```

Comment:
A docstring above still says "... else False". 

And why did you decide change False to None?

Reply:
@apollo13 ?

Reply:
It is no longer needed. Previously the code checked the return value based on this and then changed behavior. Now the proxy protocol parsing is fully "transparent" and leaves the streams in a usable state for follow up. Will adjust the comment in a follow up commit.